### PR TITLE
EBNT-378 EBNT-379

### PR DESCRIPTION
**Pull request by mike0sv**

### CODE REVIEW

File: src/ebonite/ext/numpy/dataset_source.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        return NumpyNdarrayReader(), ArtifactCollection.from_blobs(
+            {DATA_FILE: LazyBlob(lambda: save_npz(dataset.data))})
+
+
+class NumpyNdarrayReader(DatasetReader):
+    def read(self, artifacts: ArtifactCollection) -> Dataset:
+        with artifacts.blob_dict() as blobs:
+            with blobs[DATA_FILE].bytestream() as f:
+                data = np.load(f)[DATA_KEY]
+        return Dataset.from_object(data)
```
Comment (ElmiRakh) [inline src/ebonite/ext/numpy/dataset_source.py:32]: Tests

File: src/ebonite/ext/pandas/dataset_source.py [hunk 2]
```diff
@@ -1,92 +1,65 @@
...
+
+# class PandasJdbcDatasetSource(_PandasDatasetSource):
+#     def __init__(self, dataset_type: DatasetType, table: str, connection: str,
+#                  kwargs: Dict[str, Any] = None):
+#         super().__init__(dataset_type, kwargs)
+#         self.connection = connection
+#         self.table = table
+#
+#     def _read(self):
+#         return pd.read_sql_table(self.table, self.connection, **self.kwargs)
```
Comment (ElmiRakh) [inline src/ebonite/ext/pandas/dataset_source.py:65]: tests

File: src/ebonite/repository/dataset/inmemory.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from ebonite.core.objects.dataset_source import Dataset, DatasetSource, InMemoryDatasetSource
+from ebonite.repository.dataset.base import DatasetRepository
+
+
+class InMemoryDatasetRepository(DatasetRepository):
+    def save(self, dataset_id: str, dataset: Dataset) -> DatasetSource:
+        return InMemoryDatasetSource(dataset)
```
Comment (ElmiRakh) [inline src/ebonite/repository/dataset/inmemory.py:7]: tests

File: src/ebonite/repository/dataset/artifact.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class ArtifactDatasetSource(DatasetSource):
+    def __init__(self, reader: DatasetReader, artifacts: ArtifactCollection, dataset_type: DatasetType):
+        super(ArtifactDatasetSource, self).__init__(dataset_type)
+        self.reader = reader
+        self.artifacts = artifacts
+
+    def read(self) -> Dataset:
+        return self.reader.read(self.artifacts)
```
Comment (ElmiRakh) [inline src/ebonite/repository/dataset/artifact.py:31]: tests

File: src/ebonite/core/objects/core.py [hunk 5]
```diff
@@ -492,38 +561,101 @@ def delete_image(self, image: 'Image', meta_only: bool = False, cascade: bool =
...
+            raise ValueError(f'evalset {name} already in task {self}')
+        data = self._resolve_dataset(data, f'{name}_input')
+        target = self._resolve_dataset(target, f'{name}_output')
+        if not isinstance(metrics, list):
+            metrics = [metrics]
+        metrics = [self._resolve_metric(m, f'{name}_{i}') for i, m in enumerate(metrics)]
+        self.evaluation_sets[name] = EvaluationSet(data, target, metrics)
+
+    @_with_dataset
+    def add_dataset(self, name, dataset: Union[DatasetSource, AbstractDataset, Any]):
```
Comment (ElmiRakh) [inline src/ebonite/core/objects/core.py:613]: What about dataset removal? Aren't there supposed to be methods for that?

File: src/ebonite/core/objects/core.py [hunk 6]
```diff
@@ -492,38 +561,101 @@ def delete_image(self, image: 'Image', meta_only: bool = False, cascade: bool =
...
+        :param metric: Metric or raw metric object"""
+        if name in self.metrics:
+            raise ValueError(f'metric {name} already in task {self}')
         if not isinstance(metric, Metric):
             metric = MetricAnalyzer.analyze(metric)
         # TODO checks
         self.metrics[name] = metric
 
-    def evaluate_all(self):
+    def evaluate_all(self) -> Dict[str, 'EvaluationResult']:
```
Comment (ElmiRakh) [inline src/ebonite/core/objects/core.py:645]: That stuff needs tests with pipelines and models

File: src/ebonite/ext/pandas/dataset_source.py [hunk 7]
```diff
@@ -1,92 +1,227 @@
...
-                 target_col: str = None):  # TODO better target
-        super().__init__(dataset_type, kwargs, target_type, target_col)
-        self.blob = blob
-        self.format = format
+            df = super().read(file_or_path)
+
+        return df.reset_index(drop=True)
+
+
+class PandasFormatFeather(PandasFormat):
```
Comment (ElmiRakh) [inline src/ebonite/ext/pandas/dataset_source.py:157]: Tests for different formats?

Comment (ElmiRakh) [reply src/ebonite/ext/pandas/dataset_source.py:157]: False alarm

File: src/ebonite/ext/pandas/dataset.py [hunk 8]
```diff
@@ -124,31 +169,52 @@ def deserialize(self, obj):
...
         self._validate_dtypes(ret, DeserializationError)
-        return ret
+        return self.align_index(ret)
+
+    def align_types(self, df):
+        """Restores column order and casts columns to expected types"""
+        df = df[self.columns]
+        for col, expected, dtype in zip(self.columns, self.actual_dtypes, df.dtypes):
+            if expected != dtype:
+                df[col] = df[col].astype(expected)
```
Comment (ElmiRakh) [inline src/ebonite/ext/pandas/dataset.py:183]: Shouldn't we define or process error handling?

File: src/ebonite/core/objects/core.py [hunk 9]
```diff
@@ -69,6 +80,7 @@ def generate_code(self):
 
 
 class WithMetadataRepository:
+    """Intermediat abstract class for objects that can be binded to meta repository"""
```
Comment (ElmiRakh) [inline src/ebonite/core/objects/core.py:83]: Here and following - "intermediat" -> "intermediate"

File: src/ebonite/core/objects/core.py (hunk 5 above)
Comment (mike0sv) [reply src/ebonite/core/objects/core.py:613]: good one

**Pull request merged**